homeassistant/components/sensor/kwb.py

```diff
@@ -0,0 +1,113 @@
+"""
+Support for KWB Easyfire.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.kwb/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.const import (CONF_HOST, CONF_PORT, CONF_DEVICE,
+                                 CONF_NAME, EVENT_HOMEASSISTANT_STOP,
+                                 STATE_UNKNOWN)
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+import homeassistant.helpers.config_validation as cv
+
+
+REQUIREMENTS = ['pykwb==0.0.8']
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_HOST = 'localhost'
+DEFAULT_PORT = 23
+DEFAULT_TYPE = 'tcp'
+DEFAULT_RAW = False
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_NAME = 'KWB'
+
+MODE_SERIAL = 0
+MODE_TCP = 1
+
+CONF_TYPE = 'type'
+CONF_RAW = 'raw'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_TYPE, default=DEFAULT_TYPE): vol.In(['tcp', 'serial']),
+    vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
```

Comment:
I have investigated this further, and this works:

```
from voluptuous import *

SERIAL_SCHEMA = {
    Optional("raw"): str,
    Optional("name"): str,
    Required("device"): str,
    Required("type"): 'serial',
}

ETHERNET_SCHEMA = {
    Optional("raw"): str,
    Optional("name"): str,
    Required("host"): str,
    Required("port"): str,
    Required("type"): 'tcp',
}


PLATFORM_SCHEMA = Schema(
    Any(SERIAL_SCHEMA, ETHERNET_SCHEMA),
)


PLATFORM_SCHEMA({"device":"/xyz","type":"serial"})
PLATFORM_SCHEMA({"host":"test", "port":"test","type":"tcp"})

```

but in the component, the equivalent hangs on startup, yet again:
```

SERIAL_SCHEMA = {
    vol.Optional(CONF_RAW, default=DEFAULT_RAW): cv.boolean,
    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
    vol.Required(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
    vol.Required(CONF_TYPE, default=DEFAULT_TYPE): 'serial',
}

ETHERNET_SCHEMA = {
    vol.Optional(CONF_RAW, default=DEFAULT_RAW): cv.boolean,
    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
    vol.Required(CONF_HOST, default=DEFAULT_HOST): cv.string,
    vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
    vol.Required(CONF_TYPE, default=DEFAULT_TYPE): 'tcp',
}


PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    vol.Any(SERIAL_SCHEMA, ETHERNET_SCHEMA)
)
```